### ISSUE TITLE
List Secrets with recursive option.

### ISSUE THREAD

**Issue by rsankaranarayanan** [hvac/hvac#517]
List Secrets is listing only the parent directory. This options is not giving the child directory details like recursive (In linux we can use ls -r command to find out the child directory).

**Comment by rsankaranarayanan** [comment]
To address this,
I wrote small function in python. please look at this and suggest.

    def iterative(self, spath=None, parentlist=None):
        if spath:
            if parentlist is None:
                parentlist = []
            value = MyVault.list_file(self.vault_client, self.service_secret_path + "/" + spath)
            if value and isinstance(value, list):
                for v in value:
                    conpath = spath.encode('utf-8') + "/" + v
                    parentlist.append(conpath.replace("//", "/"))
                    self.iterative(conpath, parentlist)
            return parentlist
        else:
            value = MyVault.list_file(self.vault_client, self.service_secret_path)
            return value

    def vault_list_path(self, spath=None):
        if spath:
            retvalue = []
            value = self.iterative(spath)
            if value:
                for v in value:
                    if not v.endswith("/"):
                        retvalue.append(v)
                return retvalue
            else:
                print("{0} path is in-valid or {0} is file. please provide right directory instead of file.".format(spath))
        else:
            return self.iterative()

**Comment by drewmullen** [comment]
Here's a python tool I wrote to do this: https://github.com/drewmullen/vault-kv-migrate

Lmk any feedback

**Comment by rsankaranarayanan** [comment]
@drewmullen 
It looks good. 

Please look this https://github.com/rsankaranarayanan/sankar/tree/hvac-api-operations

and provide your feedback.

**Comment by drewmullen** [comment]
Are you writing entire files to vault?

https://github.com/rsankaranarayanan/sankar/blob/f7491cb12973ec31f5781b75d682a11536f2b90f/MyVault.py#L123-L135

**Comment by drewmullen** [comment]
looks like you have a lot of features written out for your specific use case. i think theres probably some things we could glean from your code if you worked it to fit into our object context and provided some documentation. specifically the oidc auth stuff. we have an [okta auth method](https://github.com/hvac/hvac/blob/develop/hvac/api/auth_methods/okta.py)right now but could use a generic oidc (im not super familiar with this topic so feel free to correct me).

Regarding the recursion method - ive thought about this a lot, as you can see i wrote my own tool because it was useful to me. Ultimately, ive never pushed it to HVAC because its not a feature of Vault and HVAC is basically a python wrapper for the Vault API.

@jeffwecan has been around this project longer than me so I'm curious what are your thoughts on providing extended features that arent native to vault (such as recursively listing a kv path).

**Comment by rsankaranarayanan** [comment]
> https://github.com/rsankaranarayanan/sankar/blob/f7491cb12973ec31f5781b75d682a11536f2b90f/MyVault.py#L123-L135

Yes. If we want to write file from local to vault. We can use "write_config_file(client, path, filepath)" function.

If we want to read the key and write into file. We can use (read_pub_file(client, vaultPath, filePath)) function.

We took MyVault.py file from open source. After that we updated few function as per needs.

**Comment by drewmullen** [comment]
after considering this, i personally feel we should not include features that arent provided by hashi. instead, in the documentation we can mention the use case and point to available tools such as https://github.com/drewmullen/vault-kv-migrate
